File: python_client/yavrha.py
#!/usr/bin/env python3
'''
               Yavrha
     Copyright (C) Carlos Silva, 2013
      (csaleman [at] gmail [dot] com)

      http://code.google.com/p/yavrha/
'''

'''
    This file is part of Yavrha.

    Yavrha is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Yavrha is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Yavrha.  If not, see <http://www.gnu.org/licenses/>.

'''
# This code connect to a serial port and send data to a MQTT server

import serial
import json
import time
import mosquitto

# Global Variables
REFRESH_DELAY = 1           # Delay in seconds to send new values only
ACTIVE_NODES = []
NODES = {}          # Nested Dictionary in {"node1":{"name":"blabla", "type":1, etc.}, "node2":{"name":"blabla", etc...
MQTT_SERVER = "test.mosquitto.org"
MQTT_TOPIC = "yavrha"	

# Serial Port Configuration
ser = serial.Serial('/dev/ttyACM0',19200, timeout=.1)


# Create Mosquitto Instance
client = mosquitto.Mosquitto("yavrha-client")
client.connect(MQTT_SERVER)

# ********************************************************
# Get Configuration data from serial port and publish it in MQTT Server
# ********************************************************

def Send_Cfg_Data():
    # Load Node Configuration
    command = bytes("print_cfg \r\n","utf-8")
    # make sure input buffer is empty
    ser.flushInput()
    ser.write(command)
    # read with the intention of delete the command sent
    ser.readline()
    Node_Cfg = (ser.read(1000))
    Node_Cfg_string = "[" + Node_Cfg.decode() +"]"
    #print(Node_Cfg_string,"\n******\n")
    Node_Cfg_Obj = json.loads(Node_Cfg_string)
    #print(Node_Cfg_Obj[0]['channel'])
    #print(Node_Cfg_Obj[0]['home_addr'])
    #print(Node_Cfg_Obj[0]['node0'])

#   This is a loop through all elements in Node_Cfg_Obj
    i=0
    while 'node'+str(i) in Node_Cfg_Obj[0]: 
#   Add active Nodes in the ACTIVE_NODES Global variable
        if Node_Cfg_Obj[0]['node'+str(i)]['enable'] == 1:
            ACTIVE_NODES.append(i)

#   Load information in NODES object
#   First check if the NODE object exist in NODES before setting properties, if not it create an emptly object 
            if 'node'+str(i) not in NODES:
                NODES["node"+str(i)] = {}     

#   Copy Nodes properties from Node_Cfg_Obj to NODES
            NODES["node"+str(i)]["name"] = str(Node_Cfg_Obj[0]['node'+str(i)]['name'])
            NODES["node"+str(i)]["type"] = str(Node_Cfg_Obj[0]['node'+str(i)]['type'])
            NODES["node"+str(i)]["address"] = str(Node_Cfg_Obj[0]['node'+str(i)]['address'])
            NODES["node"+str(i)]["number"] = str(Node_Cfg_Obj[0]['node'+str(i)]['number'])
            
#   Publish Nodes properties in MQTT Server with retain flag
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/name",str(Node_Cfg_Obj[0]['node'+str(i)]['name']), 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/type",str(Node_Cfg_Obj[0]['node'+str(i)]['type']), 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/address",str(Node_Cfg_Obj[0]['node'+str(i)]['address']), 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/number",str(Node_Cfg_Obj[0]['node'+str(i)]['number']), 1, True)

# If Node was deactivated, this publish a empty message with retain flag to delete old retained messages if any
        else:
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/name",None , 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/type",None , 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/address",None, 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(i)+"/number",None, 1, True)
        i += 1


# ********************************************************
# Get data from serial port and publish it in MQTT Server
# ********************************************************
def Send_Data():
    time.sleep(0.1)    
    ser.flushInput()    # make sure input buffer is empty
    command = bytes("get \r\n","utf-8")
    ser.write(command)
    ser.readline()          # read with the intention of delete the command sent
    Node_Data = (ser.read(500))
    Node_Data_string = "[" + Node_Data.decode() +"]"
#    print(Node_Data_string)

    try:    
        Node_Data_Obj = json.loads(Node_Data_string)
    except	ValueError:
        print("Value Error with json.loads(Node_Data_string)\n")
        print(Node_Data_string)
        return 0
# Loop through all active nodes

    for item in ACTIVE_NODES:

# This check if the key exist in the object if not it create the key, this is to avoid error in the following if
        if 'msgid' not in  NODES['node'+str(item)]:
            NODES['node'+str(item)]["msgid"] = " "

        if 'data0' not in  NODES['node'+str(item)]:
            NODES['node'+str(item)]["data0"] = " "

        if 'data1' not in  NODES['node'+str(item)]:
            NODES['node'+str(item)]["data1"] = " "

        if 'data2' not in  NODES['node'+str(item)]:
            NODES['node'+str(item)]["data2"] = " "

        if 'data3' not in  NODES['node'+str(item)]:
            NODES['node'+str(item)]["data3"] = " "

# Load new data in NODES object
# Check if new data is different that old one, if so it update the value in the NODES object and publish the value.

#        if (NODES['node'+str(item)]["msgid"] != str(Node_Data_Obj[0]['node'+str(item)]['msgid'])):        
#            NODES['node'+str(item)]["msgid"] = str(Node_Data_Obj[0]['node'+str(item)]['msgid'])
#            client.publish(MQTT_TOPIC + "/node"+str(item)+"/msgid",str(Node_Data_Obj[0]['node'+str(item)]['msgid']), 1, True)
            
        if (NODES['node'+str(item)]["data0"] != str(Node_Data_Obj[0]['node'+str(item)]['data0'])):    
            NODES['node'+str(item)]["data0"] = str(Node_Data_Obj[0]['node'+str(item)]['data0'])
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/msgid",str(Node_Data_Obj[0]['node'+str(item)]['msgid']), 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/data0",str(Node_Data_Obj[0]['node'+str(item)]['data0']), 1, True)
      
        if (NODES['node'+str(item)]["data1"] != str(Node_Data_Obj[0]['node'+str(item)]['data1'])):
            NODES['node'+str(item)]["data1"] = str(Node_Data_Obj[0]['node'+str(item)]['data1'])
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/msgid",str(Node_Data_Obj[0]['node'+str(item)]['msgid']), 1, True)            
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/data1",str(Node_Data_Obj[0]['node'+str(item)]['data1']), 1, True)
      
        if (NODES['node'+str(item)]["data2"] != str(Node_Data_Obj[0]['node'+str(item)]['data2'])):
            NODES['node'+str(item)]["data2"] = str(Node_Data_Obj[0]['node'+str(item)]['data2'])
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/msgid",str(Node_Data_Obj[0]['node'+str(item)]['msgid']), 1, True)            
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/data2",str(Node_Data_Obj[0]['node'+str(item)]['data2']), 1, True)

        if (NODES['node'+str(item)]["data3"] != str(Node_Data_Obj[0]['node'+str(item)]['data3'])):
            NODES['node'+str(item)]["data3"] = str(Node_Data_Obj[0]['node'+str(item)]['data3'])
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/msgid",str(Node_Data_Obj[0]['node'+str(item)]['msgid']), 1, True)
            client.publish(MQTT_TOPIC + "/node"+str(item)+"/data3",str(Node_Data_Obj[0]['node'+str(item)]['data3']), 1, True)

		
def Received_Cmd(msg):
    Node = str(msg.topic).lstrip(MQTT_TOPIC+'/').rstrip("/cmd").strip("node")
    command = bytes("send "+Node+" "+msg.payload.decode()+" \r\n","utf-8")
    ser.write(command)
#    print(ser.readline().decode())

def on_connect(mosq, obj, rc):
    if rc == 0:
        print("Connected successfully.")

def on_message(mosq, obj, msg):
#    print("Message received on topic "+str(msg.topic)+" with QoS "+str(msg.qos)+" and payload "+msg.payload.decode())
    Received_Cmd(msg)


# Subscribe to nodes/cmd.
client.subscribe(MQTT_TOPIC+"/+/cmd", 0)

# Set Call back functions
client.on_connect = on_connect
client.on_message = on_message

# Send Configuration Data
Send_Cfg_Data()


# Main Loop 
while True:
    start_time = int(time.time())   
    Send_Data()

# Inner loop to call mosquitto client in order to keep connection alive.
    while (int(time.time()) < start_time + REFRESH_DELAY):
        client.loop()
        

File: web_client/yavrha.js
/*
               Yavrha
     Copyright (C) Carlos Silva, 2013
      (csaleman [at] gmail [dot] com)

      http://code.google.com/p/yavrha/
*/

/*
    This file is part of Yavrha.

    Yavrha is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Yavrha is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Yavrha.  If not, see <http://www.gnu.org/licenses/>.

*/


// Global Variables
	var localStorageKey1 = "MQTTServer";
	var localStorageKey2 = "MQTTTopic";
 	var localStorageKey3 = "MQTTServerUser";
	var localStorageKey4 = "MQTTServerPassword";
	var SERVER             // Used to store Server
	var TOPIC = "";        // Used to store TOPIC
    var NodesObj = {};     // Main object where nodes data is stored
    
// MQTT Object 
    var client;


// Nodes Object
// {
//	 "node1":{  "name":"2nd_temp",
//				"type":"1",
//				"address":"230",
//				"number":"1",
//				"msgid":"83",
//				"data0":"19",
//				"data1":"1",
//				"data2":"0",
//				"data3":"0" }
//	"node2": { etc .. .. 

	

//********************************************************
// 		This is the function that initiate jquery.
//********************************************************

        $(document).on('pageinit', function() {
//		Check if local storage is suported in the browser           
			if (Modernizr.localstorage) {
                showStoredValue();
//      Save server name in Global variable				
				SERVER = localStorage.getItem(localStorageKey1)
				MQTTSetup(SERVER);
            }
            else {
                $('#message').text("Unfortunately your browser doesn't support local storage");
                $('#addToStorage').attr('disabled', 'disabled');
                $('#message').show();
            }
           
	        $('#addToStorage').click(function(e) {
//				Save values into localstorage                
				localStorage.setItem(localStorageKey1, $('#MQTTServer').val());
				localStorage.setItem(localStorageKey2, $('#MQTTTopic').val());
				localStorage.setItem(localStorageKey3, $('#MQTTServerUser').val());
				localStorage.setItem(localStorageKey4, $('#MQTTServerPassword').val());

//				Reload page with new settings.
				location.reload();
//				Show Refresh values in page.
				showStoredValue();
//				Prevent Default action. In this case submit form.
                e.preventDefault();

            });
        });

//********************************************************
// Function to get values from browser's local storage
//********************************************************
function showStoredValue() {
// MQTT Server
        	
		$('#MQTTServer').val(localStorage.getItem(localStorageKey1));
       
// MQTT Topic
	var item2 = localStorage.getItem(localStorageKey2);
        	
		$('#MQTTTopic').val(item2);

// Load Topic in global variable TOPIC		
		TOPIC = item2;           

// MQTT Server User
	var item3 = localStorage.getItem(localStorageKey3);
        
		$('#MQTTServerUser').val(item3);

// MQTT Server Password
	var item4 = localStorage.getItem(localStorageKey4);
        		
		$('#MQTTServerPassword').val(item4);
}

//********************************************************
// This function configure the MQTT instance with a server.
//********************************************************
function MQTTSetup(server)
{
   if (server != null) {
    	console.log("Trying to Connect to " + server);
    	client = new Messaging.Client( server, 80, "ClientId_"+ parseInt(Math.random() * 100, 10));
// Set callback function
    	client.onConnect = onConnect;
        client.onMessageArrived = onMessageArrived;
        client.onConnectionLost = onConnectionLost;
        client.connect({onSuccess:onConnect});

	}

}

// Test function to delete

function onMessageArrivedVoid(message) {

    console.log(message.destinationName + " - " + message.payloadString);

}

//********************************************************
// This Callback function after connection Lost
//********************************************************

function onConnectionLost() {
     
     console.log("Connection Lost");

// Try to reconnect     

   client.connect({onSuccess:onConnect});
     
  }


//********************************************************
// This Callback function after Success Connect
//********************************************************


function onConnect() {
     
     console.log("Connected");
     client.subscribe(TOPIC);
     console.log("Subscribed to " + TOPIC);
     
  }


//********************************************************
// This function Refresh the values every 5 seconds.
//********************************************************

window.setInterval(RefreshFunctions, 5000);

function RefreshFunctions() { 
    

//      Update all controls values
        UpdateValues();

}


//********************************************************
// This function is a callback from the MQTT instance
//********************************************************
function onMessageArrived(message){
  
//  Split topic to get Node Number ej. "topic/node1/name"

	TopicArray = message.destinationName.split('/');
	NodeNumber = TopicArray[1];
	MsgName	   = TopicArray[2];
	
	//console.log(NodeNumber);
	//console.log(MsgName);

//	cannot add value in an object that doesn't exist yet.
//	this create an emptly object for the node number if it doesn't exist first

	if (typeof NodesObj[NodeNumber]=='undefined'){
		
		NodesObj[NodeNumber] = {};
		NodesObj[NodeNumber][MsgName] = message.payloadString;
	}
	else {
		
//		Save the value "Payload" in the property "MsgName" for a given Node

            NodesObj[NodeNumber][MsgName] = message.payloadString;

    }


//	If message is "Number" which is the last configuration message sent by the python client call the PaintScreen function.

	if (MsgName == "number") {
		
		PaintScreen();
//  Make sure of update all controls values after PaintScreen.
        
        UpdateValues();
	}
        
//  Update Data only if a data0 to data3 is posted.

     if (MsgName == "data0" || MsgName == "data1" || MsgName == "data2" || MsgName == "data3") {    
       
         UpdateValues();
     }
 
//	 $('#DebugWindow').val(JSON.stringify(NodesObj));
//	 console.log(NodesObj);	

}

//********************************************************
// This function create  the custom forms widgets
//********************************************************
function PaintScreen()
{
//		Remove all li in listview
		$('ul#MainUl').children().remove('li');
//		Print a debug text area       
//		$('ul#MainUl').append('<li data-role="list-divider">Debug Window</li><li> <textarea rows="5" cols="25" id="DebugWindow"></textarea>');

	for (var NodeNumber in NodesObj) {

		Node = NodesObj[NodeNumber];
		
		if (Node["type"] > 127) {
			
			switch (parseInt(Node["type"])){
				// Device type 128 On/Off device
				case 128:
					$('ul#MainUl').append('<li data-role="list-divider">Switch ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li><select name="node'+ Node["number"] + '" id="node'+ Node["number"] + '" data-role="slider" data-mini="true"> \
										   <option value="0">Off</option> <option value="1">On\
										   </option> </select></li>');
					$('#node'+ Node["number"] ).on( 'slidestop', function( event ) { PostSwitchMessage($(this).attr('id') ,$(this).val()); });		
                    break;
            	
                // Device type 129 On/Off device with Dimmer		
            	case 129:
					$('ul#MainUl').append('<li data-role="list-divider">Dimmer ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li><select name="Snode'+ Node["number"] + '" id="Snode'+ Node["number"] + '" data-role="slider" data-mini="true"> \
										   <option value="0">Off</option> <option value="1">On\
										   </option> </select>\
                                           <label for="node'+ Node["number"] + '" class="ui-hidden-accessible">Input slider:</label> \
                                           <input type="range" name="node'+ Node["number"] + '" id="node'+ Node["number"] + '" \
                                           value="5" min="0" max="10" step="1" data-highlight="true" data-mini="true"/></li>');
                    
//                	Refresh the CSS in the list. The event handlers MUST be set after refreshing the list. 
                    $('#MainUl').listview("refresh").trigger("create"); 
//                  Set event handler to handle dimmer                                       
                    $('#node'+ Node["number"]).ready( CreateDimmerEvent(Node["number"]));
//                  Set event handler to hande switch
                    $('#Snode'+ Node["number"]).ready(  CreateDimmerSwitchEvent(Node["number"]));
                 
                    break;
                
                // Device type 130 Servo Device 				
                case 130:
					$('ul#MainUl').append('<li data-role="list-divider">Servo ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li> <input type="text" id="Node1" name="Node1"/> </li>');
					break;
					

			}
		} 
	}
		
	for (var NodeNumber in NodesObj) {
		
		Node = NodesObj[NodeNumber];
		
		if (Node["type"] < 128) {
			
			switch (parseInt(Node["type"])){
				
				case 1:
					$('ul#MainUl').append('<li data-role="list-divider">Temperature ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li id="node'+ Node["number"]+'" >00.0</li>');
					break;
				case 2:
					$('ul#MainUl').append('<li data-role="list-divider">Status ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li> <input type="text" id="Node1" name="Node1"/> </li>');
					break;
				case 3:
					$('ul#MainUl').append('<li data-role="list-divider">Power ' + Node["name"] + ' </li>');
			 		$('ul#MainUl').append('<li> <input type="text" id="Node1" name="Node1"/> </li>');
					break;
					
			}
		}
	}



// 	This refresh the CSS in the list.
//	This works in JQ Mobile 1.3+
//	$('#MainUl').listview("refresh");
// 	This works in all JQ Mobile Version 	 
	$('#MainUl').listview("refresh").trigger("create");
}
//********************************************************
// This function create event handlers for Dimmer Slider
// 
//********************************************************
function CreateDimmerEvent(NodeNumber){
  
// The following check if the handler exist before creating one.
    var ev = $._data($('#node'+ NodeNumber)[0], 'events');

    if (!(ev && ev.slidestop)) {
//  Create a event handler        
            $('#node'+ NodeNumber).on('slidestop', function( event ) { PostDimmerMessage($(this).attr('id'));  });
        }

}

//********************************************************
// This function create event handlers for Dimmer Switch
//********************************************************
function CreateDimmerSwitchEvent(NodeNumber){
  
// The following check if the handler exist before creating one.
    var ev = $._data($('#Snode'+ NodeNumber)[0], 'events');

    if (!(ev && ev.slidestop)) {
//  Create a event handler        
            $('#Snode'+ NodeNumber).on('slidestop', function( event ) { PostDimmerMessage('node'+ NodeNumber); });
        }

}


//********************************************************
// Function to update values
//********************************************************
function UpdateValues(){

		for (var NodeNumber in NodesObj) {
						
			Node = NodesObj[NodeNumber];

			switch (parseInt(Node["type"])){
				
				case 1:
					var sign = (Node["data1"] = 1) ? "" : "-";
					var tempC = parseInt(sign + Node["data0"] );
					var tempF = Math.round((tempC*1.8)+32);
					$('#'+NodeNumber).text(tempC + "  \u00B0C ( " + tempF + " \u00B0F )"  );	
					break;
				case 2:
					
					break;
				case 3:
					
					break;
				case 128:
                    // This command assign a new value to the widget and refresh it.                        
                        $('#'+NodeNumber).val(Node["data0"]).slider("refresh");
					break;
				case 129:
                    // This command assign a new value to the widget and refresh it.                        
                        $('#S'+NodeNumber).val(Node["data0"]).slider("refresh");
                        $('#'+NodeNumber).val(parseInt(Node["data1"]) / 25).slider("refresh");

					break;
				case 130:
					
					break;


			}
		}
}

  
//***************************************************************************************
// Function to Post Messages back to MQTT server from Device Type 128 On/OFF Switch
//***************************************************************************************
function PostSwitchMessage(Node, PostVal)
{
    postTopic = TOPIC.substring(0, TOPIC.length - 1) + Node + "/cmd";    
    message = new Messaging.Message(PostVal);
    message.destinationName = postTopic;
    client.send(message);

	//console.log(postTopic);
	//console.log(PostVal);
}

//*********************************************************************************************
// Function to Post Messages back to MQTT server from Device Type 129 On/OFF Switch with dimmer
// arguments: node
//  Function will look for values in Snode for switch on/off and in node for dimmer pwm
//*********************************************************************************************
function PostDimmerMessage(Node)
{
   //Since slider will step from 1 to 10, I will multiply the value by 25
    var newPostVal = parseInt( $('#'+Node).val()) * 25;

    postTopic = TOPIC.substring(0, TOPIC.length - 1) + Node + "/cmd";
    message = new Messaging.Message($('#S'+Node).val() + " " + newPostVal);
    message.destinationName = postTopic;
    client.send(message);

	//alert(postTopic +" "+ $('#S'+Node).val() + " " + newPostVal);
}
